### PR TITLE
Save network interactions to allow tests to run offline

### LINKED ISSUE #234: tests should not need internet conection
The testsuite should not depend on the status of the internet connection. Coding on trains and planes is fun and should be supported :)

## Expected Behavior
It is a common practice to mock the external requests using a library like `mock` [doc](https://docs.python.org/3/library/unittest.mock.html). The calls to remote APIs are replaced by reads to local files.

## Possible Solution
I did [something like that in the past](https://github.com/PatchPorting/patcher/blob/master/tests/remote_tests.py) and it could be useful initially (not sure if my way is totally correct neither).

## Steps to Reproduce
1. Put your computer offline
2. Run the tests
3. Enjoy seeing them failing.

### PR DESCRIPTION
**Pull request by 1ucian0**
### Summary

Following the discussion in #234, we would like to have a way to run tests offline. These are some of the advantages:
 - It's possible to contribute to QISKit while traveling on a train
 - Online tests are expensive in terms of time and bandwidth
 - Statistics about executions are inflated because tests are hitting the real API
 - Run the test should not require a QE token
 - Tests are deterministic and do not depend on the state of the server (eg if a particular backend is offline temporarily)

This PR tackles the issue by integrating with [VCRpy](https://vcrpy.readthedocs.io/en/latest/index.html), a well-maintained library to record and replay HTTP interactions. Working cassettes are also provided in this PR.

### Details and comments

Fixes #234 and makes #603 unnecessary 

A new make rule `make test_record` was added to re-record the cassettes.

### CODE REVIEW

File: test/python/common.py [hunk 1]
```diff
@@ -289,5 +298,200 @@ def _is_ci_fork_pull_request():
...
-SKIP_ONLINE_TESTS = os.getenv('SKIP_ONLINE_TESTS', _is_ci_fork_pull_request())
 SKIP_SLOW_TESTS = os.getenv('SKIP_SLOW_TESTS', True) not in ['false', 'False', '-1']
+RECORD_TEST_RESPONSE = os.getenv('RECORD_TEST_RESPONSE', False) is not False
+VCR_MODE = 'none'
+if RECORD_TEST_RESPONSE:
+    SKIP_SLOW_TESTS = True  # TODO Activate later
+    VCR_MODE = 'all'
+
+
+class IdRemoverPersister(FilesystemPersister):
```
Comment (diego-plan9) [inline test/python/common.py]: Not a full review yet - but can you move this class and the VCR related functionality (except the bits you might need in the decorator, etc) to a new file, in the hopes of keeping `common.py` at a manageable size and complexity?

**Comment by diego-plan9** [comment]
Before reviewing the implementation, I think we need to address the change in the default behaviour that you are introducing along with the changes: whether, by default, the online tests should be mocked or not.

The points you mention in the PR description are indeed valid and in general desirable, and it seems we will very likely be able address them - but I think other considerations need to be thrown into the decision. In particular:
* the whole purpose of unit testing is to have the highest assurances that the software is working as expected. In our case, they can only ultimately be marked as passed when they are executed against the real API.
* mocking is not guaranteed to be 100% 1:1 with the API, both in theory and in practice.

and personally I think they should be given quite a lot of weight  in the decision - I'm all in for smoothening all rough edges for contributors and address practical concerns (as the root of #603 ), but not at the cost of working in the opposite direction of the first point.

In particular, I think there is a very real risk of the mocking getting outdated (not reflecting the real API), and going unnoticed, rippling up in a number of side effects: developers need to explicitly opt-in for making sure their test ultimately pass, they need to be aware of the mocking internals, and in general I think it ends up complicating the flow for them and for us, introducing new potential points of failure. If mocking is opt-in, there is basically just one thing developers need to do before submitting a PR: running the tests, same as every other Python project. The ones willing to use the mock and make a voluntary trade-off of "time vs assurances" can happily do so (and probably should!), in the same way they can run a single test or have their preferred flow of choice (as a side note, I'd personally like to keep the ability of skipping the online tests altogether, expanding on the "groups of tests" concept that we implicitly have already, but as part of another PR): giving them the mechanisms and leaving the decision to use them seems a good approach to me.

TL;DR: what about something like:
* check if there are credentials (addressing #603)
   * if they don't, print a warning, and run the mocked tests
   * if they do:
      * run the mocked test if the user opts-in (via environment variable or similar)
      * otherwise, run the default, non-mocked tests

File: test/python/common.py (hunk 1 above)
Comment (1ucian0) [reply test/python/common.py]: Addressed in b3215c8

**Comment by 1ucian0** [comment]
ToDo:

- [X] Rename to WIP
- [ ] Implement the logic that @diego-plan9 suggested.
- [ ] Convert doc from md to rst
- [ ] Convert docstrings to Google style

Feel free to add things here.

**Comment by 1ucian0** [comment]
Fix #747

File: test/python/http_recorder.py [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+import json
+from vcr.persisters.filesystem import FilesystemPersister
+from vcr import VCR
+
+
+class IdRemoverPersister(FilesystemPersister):
+    """
+    IdRemoverPersister is a VCR persister. This is, it implements a way to save and load cassettes.
+    This persister in particular inherits load_cassette from FilesystemPersister (basically, it
+    loads a standard cassette in the standard way from the FS). On the saving side, it removes some
```
Comment (delapuente) [inline test/python/http_recorder.py]: Nit: `FS` -> `files system`; `it removes some fields in the JSON content of the responses for dummy values.` -> `it replaces some fields in the JSON content of the responses with dummy values`.

File: test/python/http_recorder.py [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+        Filters the requests from cassette_dict
+
+        Args:
+            string2find (str): request path
+            cassette_dict (dict): a VCR cassette dictionary
+
+        Returns:
+            Request: VCR's representation of a request.
+        """
+        requests_indeces = [i for i, x in enumerate(cassette_dict['requests']) if
```
Comment (delapuente) [inline test/python/http_recorder.py]: Nit: `request_indeces` -> `request_indices`.

File: test/python/http_recorder.py [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+class IdRemoverPersister(FilesystemPersister):
+    """
+    IdRemoverPersister is a VCR persister. This is, it implements a way to save and load cassettes.
+    This persister in particular inherits load_cassette from FilesystemPersister (basically, it
+    loads a standard cassette in the standard way from the FS). On the saving side, it removes some
+    fields in the JSON content of the responses for dummy values.
+    """
+
+    @staticmethod
+    def get_responses_with(string2find, cassette_dict):
```
Comment (delapuente) [inline test/python/http_recorder.py]: Non-blocking but these auxiliary static methods could be auxiliary module-level private functions.

Comment (delapuente) [reply test/python/http_recorder.py]: Nit: `string2find` -> `string_to_find`.

File: test/python/http_recorder.py (hunk 3 above)
Comment (delapuente) [reply test/python/http_recorder.py]: What `x` stands for? Can you make the variable name more descriptive?

File: test/python/http_recorder.py [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+
+        Returns:
+            Request: VCR's representation of a request.
+        """
+        requests_indeces = [i for i, x in enumerate(cassette_dict['requests']) if
+                            string2find in x.path]
+        return [cassette_dict['responses'][i] for i in requests_indeces]
+
+    @staticmethod
+    def get_new_id(field, path, id_tracker, _type=str):
```
Comment (delapuente) [inline test/python/http_recorder.py]: According to PEP8, it is preferable if you use `type_`.

File: test/python/http_recorder.py [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+    @staticmethod
+    def get_maching_dicts(data_dict, map_list):
+        """
+        Find subdicts that are described in map_list.
+        Args:
+            data_dict (dict): in which the map_list is going to be searched.
+            map_list (list): the list of nested keys to find in the data_dict
+
+        Returns:
+            dict: the dictionary in which matches map_list.
```
Comment (delapuente) [inline test/python/http_recorder.py]: According to the code, this is returning a list. Can you update the docstring accordingly?

File: test/python/http_recorder.py [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+                                   'qasms.result.data.additionalData.seed'],
+                      'api/Backends/ibmqx5/queue/status': ['lengthQueue'],
+                      'api/Backends/ibmqx4/queue/status': ['lengthQueue']}
+        IdRemoverPersister.remove_ids(ids2remove, cassette_dict)
+        super(IdRemoverPersister, IdRemoverPersister).save_cassette(cassette_path,
+                                                                    cassette_dict,
+                                                                    serializer)
+
+
+def purge_headers(headers):
```
Comment (delapuente) [inline test/python/http_recorder.py]: Can you convert `purge_headers` and `unordered_query_matcher` into private functions and place them at the end of the file, leaving `http_recorder` (which is the most important) in the first place?

File: test/python/http_recorder.py [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+    Remove headers from the response.
+
+    Args:
+        headers (list): headers to remove from the response
+
+    Returns:
+        callable: for been used in before_record_response VCR constructor.
+    """
+
+    header_list = list()
```
Comment (delapuente) [inline test/python/http_recorder.py]: Use `[]` instad of `list()`.

File: test/python/http_recorder.py [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+    for item in headers:
+        if not isinstance(item, tuple):
+            item = (item, None)
+        header_list.append((item[0], item[1]))
```
Comment (delapuente) [inline test/python/http_recorder.py]: Did you wanted to add a copy of the tuple? If not, you can simply add `ítem`.

File: test/python/http_recorder.py [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+            except KeyError:
+                pass
+        return response
+
+    return before_record_response
+
+
+def unordered_query_matcher(request1, request2):
+    """
+    A VCR matcher (a la VCR.matcher) that ignores the order of the query strings and ther
```
Comment (delapuente) [inline test/python/http_recorder.py]: It is `query string` (notice the singular); also replace `ther` with `its`.

File: test/python/http_recorder.py [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+    dict1 = dict(request1.query)
+    dict2 = dict(request2.query)
+
+    if dict1 == dict2:
+        return True
+
+    if dict1.keys() != dict2.keys():
+        return False
+
+    print(dict1, '==', dict2)
```
Comment (delapuente) [inline test/python/http_recorder.py]: Remove this line.

File: test/python/http_recorder.py [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
+
+    for key, value in dict1.items():
+        try:
+            dict1[key] = json.loads(value)
```
Comment (delapuente) [inline test/python/http_recorder.py]: Do we expect to receive JSON in the querystring?

File: test/python/http_recorder.py [hunk 13]
```diff
@@ -0,0 +1,285 @@
...
+        Purge headers from response.
+
+        Args:
+            response (dict): a VCR response
+
+        Returns:
+            dict: a VCR response
+        """
+        for (header, value) in header_list:
+            try:
```
Comment (delapuente) [inline test/python/http_recorder.py]: Use [`contextlib.supress(KeyError)`](
https://docs.python.org/3/library/contextlib.html#contextlib.suppress) to make it explicit that you want to ignore this error.

File: test/python/http_recorder.py [hunk 14]
```diff
@@ -0,0 +1,285 @@
...
+    if dict1 == dict2:
+        return True
+
+    if dict1.keys() != dict2.keys():
+        return False
+
+    print(dict1, '==', dict2)
+
+    for key, value in dict1.items():
+        try:
```
Comment (delapuente) [inline test/python/http_recorder.py]: Same here.

File: test/python/common.py [hunk 15]
```diff
@@ -1,6 +1,6 @@
 # -*- coding: utf-8 -*-
 
-# Copyright 2017, IBM.
+# Copyright 2018, IBM.
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply test/python/common.py]: b8344ff

File: test/python/common.py [hunk 16]
```diff
@@ -312,46 +354,34 @@ def requires_qe_access(func):
...
         _wrapper._DEFAULT_PROVIDER = DefaultQISKitProvider()
 
-        if os.getenv('USE_ALTERNATE_ENV_CREDENTIALS', False):
-            # Special case: instead of using the standard credentials mechanism,
-            # load them from different environment variables. This assumes they
-            # will always be in place, as is used by the Travis setup.
+        if TEST_OPTIONS['skip_online']:
+            raise unittest.SkipTest('Skipping online tests')
+
+        if get_credentials(args, kwargs):
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply test/python/common.py]: 12dda56

File: test/python/common.py [hunk 17]
```diff
@@ -312,46 +354,34 @@ def requires_qe_access(func):
...
+            if TEST_OPTIONS['rec'] or TEST_OPTIONS['mock_online']:
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply test/python/common.py]: if you are recording (`TEST_OPTIONS['rec']`) or you are mocking (`TEST_OPTIONS['mock_online']`), you need to decorate the test function. Otherwise, you go as currently.

File: test/python/common.py [hunk 18]
```diff
@@ -312,46 +354,34 @@ def requires_qe_access(func):
 
     @functools.wraps(func)
     def _(*args, **kwargs):
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply test/python/common.py]: 1644c03

File: test/python/common.py [hunk 19]
```diff
@@ -376,5 +406,74 @@ def _is_ci_fork_pull_request():
...
+        defaults[opt] = if_true[opt]()
+    return defaults
+
+
+TEST_OPTIONS = get_test_options()
+VCR_MODE = 'none'
+if TEST_OPTIONS['rec']:
+    VCR_MODE = 'all'
+
+VCR = http_recorder(VCR_MODE)
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply test/python/common.py]: 5dae969 and 37b12a9

File: Makefile [hunk 20]
```diff
@@ -29,6 +29,13 @@ style:
 test:
 	python3 -m unittest discover -s test -v
 
+test_mock:
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply Makefile]: The rule provides an easy way to opt-in. The documentation and examples point to `make test`.

File: Makefile [hunk 21]
```diff
@@ -29,6 +29,13 @@ style:
...
+	env QISKIT_TESTS=mock_online python3 -m unittest discover -s test -v
+
+test_record:
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply Makefile]: da8edaf

File: .github/CONTRIBUTING.rst [hunk 22]
```diff
@@ -205,8 +205,9 @@ Windows:
     C:\..\> set LOG_LEVEL="INFO"
     C:\..\> python -m unittest test/python/test_apps.py
 
-Additionally, an environment variable ``SKIP_ONLINE_TESTS`` can be used for
-toggling the execution of the tests that require network access to the API.
+By default, the tests run against a local copy of the HTTP interactions.
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply .github/CONTRIBUTING.rst]: 571264a

File: test/python/test_ibmqjob.py [hunk 23]
```diff
@@ -44,9 +43,7 @@ class TestIBMQJob(JobTestCase):
     Test ibmqjob module.
     """
 
-    @requires_qe_access
-    def setUp(self, QE_TOKEN, QE_URL, hub=None, group=None, project=None):
-        # pylint: disable=arguments-differ
+    def setUp(self):
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply test/python/test_ibmqjob.py]: Because the cassette auto-naming is based on the test name. Otherwise, all the `setUp` interactions are saves in the same file and might collide.

File: test/python/test_ibmqjob.py [hunk 24]
```diff
@@ -214,34 +218,42 @@ def test_run_async_device(self):
         job_ids = [job.id for job in job_array]
         self.assertEqual(sorted(job_ids), sorted(list(set(job_ids))))
 
+    @requires_qe_access
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply test/python/test_ibmqjob.py]: 3650711

File: test/python/http_recorder.py (hunk 2 above)
Comment (1ucian0) [reply test/python/http_recorder.py]: 875ca85

File: test/python/http_recorder.py (hunk 3 above)
Comment (1ucian0) [reply test/python/http_recorder.py]: 6b34604

File: test/python/http_recorder.py (hunk 4 above)
Comment (1ucian0) [reply test/python/http_recorder.py]: 8d67900

Comment (1ucian0) [reply test/python/http_recorder.py]: about the static method. What module-level private is better?

File: test/python/http_recorder.py (hunk 5 above)
Comment (1ucian0) [reply test/python/http_recorder.py]: d839c20

File: test/python/http_recorder.py (hunk 6 above)
Comment (1ucian0) [reply test/python/http_recorder.py]: 5e7bf25

File: test/python/http_recorder.py (hunk 7 above)
Comment (1ucian0) [reply test/python/http_recorder.py]: f868587

File: test/python/http_recorder.py (hunk 8 above)
Comment (1ucian0) [reply test/python/http_recorder.py]: d7cc3af

File: test/python/http_recorder.py (hunk 9 above)
Comment (1ucian0) [reply test/python/http_recorder.py]: `header_list` is consumed as `(header, value) in header_list`. Some items are not tuples, I want them to have `None` as value.

File: test/python/http_recorder.py (hunk 11 above)
Comment (1ucian0) [reply test/python/http_recorder.py]: ups... fc60410

File: test/python/http_recorder.py (hunk 13 above)
Comment (1ucian0) [reply test/python/http_recorder.py]: 2e24709

File: test/python/http_recorder.py (hunk 14 above)
Comment (1ucian0) [reply test/python/http_recorder.py]: 2e24709

File: test/python/common.py [hunk 25]
```diff
@@ -312,46 +354,34 @@ def requires_qe_access(func):
...
+                return VCR.use_cassette()(func)(*args, **kwargs)
+            return func(*args, **kwargs)
+
+        # No user credentials were found.
+        if TEST_OPTIONS['rec']:
+            raise Exception(
+                'Could not locate valid credentials. You need them for performing '
+                'tests against the remote API. Use QISKIT_TESTS=skip_online to skip this test.')
+        else:
+            args[0].log.warning("No user credentials were detected. Running with mocked data.")
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply test/python/common.py]: It was there from before. Can we do it later?

File: doc/dev_introduction.rst [hunk 26]
```diff
@@ -195,7 +195,22 @@ Windows:
     C:\..\> set LOG_LEVEL="INFO"
     C:\..\> python -m unittest test/python/test_apps.py
 
-Additionally, an environment variable ``SKIP_ONLINE_TESTS`` can be used for
-toggling the execution of the tests that require network access to the API and
-``SKIP_SLOW_TESTS`` can be used to toggling execution of tests that are
-particularly slow (default is ``True``).
+Testing options
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply doc/dev_introduction.rst]: Extracting this back is a bit complicated now...

File: test/python/common.py [hunk 27]
```diff
@@ -254,13 +252,48 @@ def slow_test(func):
...
+    def _wrapper(*args, **kwargs):
+        if TEST_OPTIONS['skip_slow']:
             raise unittest.SkipTest('Skipping slow tests')
         return func(*args, **kwargs)
 
-    return _
+    return _wrapper
+
 
+def _add_credentials(args, kwargs):
```
Comment (diego-plan9) [inline test/python/common.py]: Would it make sense to tweak this function in order to provide more clarity and easier following of the logic as a whole? As things stand, I believe there are a couple of issues that are slightly convoluted (touching a bit on other comments by @delapuente, probably):
* this function seems to be called even if the mocking is enabled, which seems counter-intuitive (and with less importance, it is called unnecessarily as a side effect):
```python
    if _add_credentials(args, kwargs):
            if TEST_OPTIONS['rec'] or ****TEST_OPTIONS['mock_online']****:
                return VCR.use_cassette()(func)(*args, **kwargs)
            return func(*args, **kwargs)
```
* the fake credentials are appended in `requires_qe_access()` - making this function no longer the only place where credentials can be added.

What about making this function the "real only place where credentials for the tests are handled"? I'd personally also try to make return a dict with the credentials instead of a boolean, mostly for avoiding to modify an argument in-place and also for making it agnostic to kwargs, but not a super strong preference. This way it can be used as:

```
def requires_qe_access():
    ...
    kwargs.update(_add_credentials(mock_online))  # warning and exception raised inside
    ...
```

Which would in turn convert `requires_qe_access()` into mostly "decide which wrapper function should be returned based on the test options". As an aside, it might also imply reverting the name change for the `_add_credentials` function :grin:.

File: requirements.txt [hunk 28]
```diff
@@ -6,3 +6,4 @@ ply>=3.10
 scipy>=0.19
 sympy>=1.0
 pillow>=4.2.1
+vcrpy
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply requirements.txt]: b33704f

File: test/python/common.py [hunk 29]
```diff
@@ -312,42 +345,34 @@ def requires_qe_access(func):
...
+                'Could not locate valid credentials. You need them for performing '
+                'tests against the remote API. Use QISKIT_TESTS=skip_online to skip this test.')
+        else:
+            args[0].log.warning("No user credentials were detected. Running with mocked data.")
             kwargs.update({
-                'QE_TOKEN': os.getenv('IBMQ_TOKEN'),
-                'QE_URL': os.getenv('IBMQ_URL')
+                'QE_TOKEN': 'dummyapiusersloginWithTokenid01',
+                'QE_URL': 'https://quantumexperience.ng.bluemix.net/api',
+                'hub': None,
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply test/python/common.py]: You mean #740? When I merge from master I missed this. Fixed in 70ecb21.

File: test/python/http_recorder.py [hunk 30]
```diff
@@ -0,0 +1,280 @@
...
+        callable: for been used in before_record_response VCR constructor.
+    """
+
+    header_list = []
+    for item in headers:
+        if not isinstance(item, tuple):
+            item = (item, None)
+        header_list.append((item[0], item[1]))
+
+    def before_record_response(response):
```
Comment (delapuente) [inline test/python/http_recorder.py]: Can you add the suffix `_callback` or `_cb` to make it clear this is a callback?

File: test/python/http_recorder.py [hunk 31]
```diff
@@ -0,0 +1,280 @@
...
+                                               'Content-Security-Policy',
+                                               'X-Content-Security-Policy',
+                                               'X-Webkit-Csp',
+                                               'content-length']))
+    my_vcr.register_matcher('unordered_query', _unordered_query_matcher)
+    my_vcr.register_persister(IdRemoverPersister)
+    return my_vcr
+
+
+def _purge_headers(headers):
```
Comment (delapuente) [inline test/python/http_recorder.py]: Can you add the suffix `_callback` here too to make it clear that the function returns a callback?

File: test/python/http_recorder.py [hunk 32]
```diff
@@ -0,0 +1,280 @@
...
+            ids2remove (dict): {request_path: [json_fields]}
+            cassette_dict (dict): a VCR cassette dictionary.
+        """
+
+        id_tracker = {}  # {old_id: new_id}
+        for path, fields in ids2remove.items():
+            responses = IdRemoverPersister.get_responses_with(path, cassette_dict)
+            for response in responses:
+                IdRemoverPersister.remove_ids_in_a_response(response, fields, path, id_tracker)
+        for old_id, new_id in id_tracker.items():
```
Comment (delapuente) [inline test/python/http_recorder.py]: I understand you use the tracker to make the same in the URIs after replacing the proper values in the response. If I'm correct, indicate it in a comment or move this into a specific method.

File: test/python/http_recorder.py [hunk 33]
```diff
@@ -0,0 +1,280 @@
...
+            if not isinstance(old_id, str):
```
Comment (delapuente) [inline test/python/http_recorder.py]: Non-blocking: making this test positive and putting the remaining block inside make it clear you want to do replace only textual values.

File: test/python/http_recorder.py [hunk 34]
```diff
@@ -0,0 +1,280 @@
...
+
+        Args:
+            jsonobj (dict): json dictionary from the response body
+            field (str): string with the field in the response to by replaced
+            path (str): request path
+            id_tracker (dict): a dictionary of the ids already assigned.
+        """
+
+        map_list = field.split('.')
+        for maching_dict in IdRemoverPersister.get_maching_dicts(jsonobj, map_list):
```
Comment (delapuente) [inline test/python/http_recorder.py]: Typo: it is `matching_dict` and `get_matching_dicts` (notice the `t` in `matching`).

File: test/python/http_recorder.py [hunk 35]
```diff
@@ -0,0 +1,280 @@
...
+            map_list (list): the list of nested keys to find in the data_dict
+
+        Returns:
+            list: a list of dictionaries, each of them matches map_list.
+        """
+        ret = []
+        if not map_list:
+            return ret
+        if isinstance(data_dict, list):
+            _ = [ret.extend(IdRemoverPersister.get_maching_dicts(i, map_list)) for i in data_dict]
```
Comment (delapuente) [inline test/python/http_recorder.py]: Convert this into a loop, please. It is easier to read and name `i` in a meaningful way.

File: test/python/common.py [hunk 36]
```diff
@@ -308,42 +356,20 @@ def requires_qe_access(func):
...
-            # Special case: instead of using the standard credentials mechanism,
-            # load them from different environment variables. This assumes they
-            # will always be in place, as is used by the Travis setup.
-            kwargs.update({
-                'qe_token': os.getenv('IBMQ_TOKEN'),
-                'qe_url': os.getenv('IBMQ_URL')
-            })
-            args[0].using_ibmq_credentials = True
+        if TEST_OPTIONS['rec'] or TEST_OPTIONS['mock_online']:
+            wrapper = VCR.use_cassette()(func)
```
Comment (delapuente) [inline test/python/common.py]: Replace with `func = VCR.use_cassette()(func)` and remove the `else` branch.

File: test/python/common.py [hunk 37]
```diff
@@ -308,42 +356,20 @@ def requires_qe_access(func):
...
 
     @functools.wraps(func)
-    def _wrapper(*args, **kwargs):
-        if SKIP_ONLINE_TESTS:
+    def _wrapper(self, *args, **kwargs):
+        if TEST_OPTIONS['skip_online']:
             raise unittest.SkipTest('Skipping online tests')
 
         # Cleanup the credentials, as this file is shared by the tests.
         from qiskit.wrapper import _wrapper
```
Comment (delapuente) [inline test/python/common.py]: Import the `_wrapper` with a different name to avoid name collision:
```python
from qiskit.wrapper import _wrapper as qiskit_wrapper
```

File: test/python/common.py (hunk 36 above)
Comment (delapuente) [reply test/python/common.py]: It seems counterintuitive we `use_casette` while recording, but it is not our fault if they chose that API, can you clarify in a comment that, when using `rec`, that implies `VCR` is in recording mode?

File: test/python/common.py [hunk 38]
```diff
@@ -368,5 +394,77 @@ def _is_ci_fork_pull_request():
     return False
 
 
-SKIP_ONLINE_TESTS = os.getenv('SKIP_ONLINE_TESTS', _is_ci_fork_pull_request())
-SKIP_SLOW_TESTS = os.getenv('SKIP_SLOW_TESTS', True) not in ['false', 'False', '-1']
+def get_test_options(option_var='QISKIT_TESTS'):
```
Comment (delapuente) [inline test/python/common.py]: This function is in charge of parsing and resolving implications between the test options. It could be better implemented in a different module or in its own class. While isolating, let's try to fix the following issues.

File: test/python/common.py [hunk 39]
```diff
@@ -368,5 +394,77 @@ def _is_ci_fork_pull_request():
...
+    """
+    Reads option_var from env and returns a dict in which the test options are set
+
+    Args:
+        option_var (str): The env var to read. Default: 'QISKIT_TESTS'
+
+    Returns:
+        dict: A dictionary with the format {<option>: (bool)<activated>}.
+    """
+    defaults = {
```
Comment (delapuente) [inline test/python/common.py]: These defaults are contradictory, not passing any flag would result in an option set which is inconsistent since `skip_online` and `run_online` would be `False` for both. 

That said, notice we have three groups of tests: `offline`, `online`, `slow`. These can run or be skipped, which implies 8 possibilities to be controlled by 3 booleans at most. But `offline` tests are always run so 2 booleans, one set to run by default (`online`), the other to skip (`slow`).

I suggest removing `run_online` and `skip_slow`. This will simplify the implication map increasing readability and maintainability.

File: test/python/common.py [hunk 40]
```diff
@@ -368,5 +394,77 @@ def _is_ci_fork_pull_request():
...
+
+        Returns:
+            bool: True, returns always True.
+        """
+
+        defaults[option] = False
+        return True
+
+    if_true = {
+        'skip_online': (lambda: turn_false('run_online') and turn_false('rec')),
```
Comment (delapuente) [inline test/python/common.py]: You can remove outer parentheses wrapping lambdas.

File: test/python/common.py [hunk 41]
```diff
@@ -368,5 +394,77 @@ def _is_ci_fork_pull_request():
...
+        'run_online': (lambda: turn_false('skip_online')),
+        'mock_online': (lambda: turn_true('run_online') and turn_false('skip_online')),
+        'skip_slow': (lambda: turn_false('run_online')),
```
Comment (delapuente) [inline test/python/common.py]: Consider the test [`test_compile_run_remote`](https://github.com/1ucian0/qiskit-terra/blob/d2809ef263cb0c7283c33991660b5659de4becc8/test/python/test_compiler.py#L200). It is online but not marked as `slow_test` (this happen a lot in the codebase) so this implication is not correct.

File: test/python/common.py [hunk 42]
```diff
@@ -368,5 +394,77 @@ def _is_ci_fork_pull_request():
...
+            'run_slow'))
+    }
+
+    opt_string = os.getenv(option_var, False)
+    if not opt_string:
+        return defaults
+
+    for opt in opt_string.split(','):
+        # This means, set the opt to True and flip all the opts that need to be rewritten.
+        defaults[opt] = if_true[opt]()
```
Comment (delapuente) [inline test/python/common.py]: Instead of explaining what is this doing, why not putting inside a function with a meaningful name?

File: test/python/http_recorder.py (hunk 30 above)
Comment (1ucian0) [reply test/python/http_recorder.py]: b0b1cc6

File: test/python/http_recorder.py (hunk 31 above)
Comment (1ucian0) [reply test/python/http_recorder.py]: b0b1cc6

File: test/python/http_recorder.py (hunk 35 above)
Comment (1ucian0) [reply test/python/http_recorder.py]: 9eebae8

File: test/python/common.py (hunk 36 above)
Comment (1ucian0) [reply test/python/common.py]: The statement `func = VCR.use_cassette()(func)` creates a funny situation in the scope static detection for lint. The error `Using variable 'func' before assignment (used-before-assignment)` pops up. Shall I ignore it?

In the way I see, you still need a cassette to record in a real-life VCR. Comment in: 71a8cb6

File: test/python/common.py (hunk 37 above)
Comment (1ucian0) [reply test/python/common.py]: a49e769

File: test/python/common.py (hunk 38 above)
Comment (1ucian0) [reply test/python/common.py]: b29c093

File: test/python/common.py (hunk 39 above)
Comment (1ucian0) [reply test/python/common.py]: e6a5cd8

File: test/python/common.py (hunk 40 above)
Comment (1ucian0) [reply test/python/common.py]: aed183a

File: test/python/common.py (hunk 41 above)
Comment (1ucian0) [reply test/python/common.py]: removed with e6a5cd8

File: test/python/common.py (hunk 42 above)
Comment (1ucian0) [reply test/python/common.py]: f427e70

File: test/python/common.py (hunk 39 above)
Comment (1ucian0) [reply test/python/common.py]: `run_online` and `skip_online` are not exactly opposite. `run_online` forces the use network against the real API. I agree that the way to have that options is by having credentials.

File: test/python/test_ibmqjob.py [hunk 43]
```diff
@@ -57,11 +53,11 @@ def setUp(self, qe_token, qe_url):
         qc.cx(qr[0], qr[1])
         qc.measure(qr, cr)
         self._qc = qc
-        self._provider = IBMQProvider(qe_token, qe_url)
-        self._local_provider = LocalProvider()
```
Comment (1ucian0) [inline test/python/test_ibmqjob.py]: This line was inserted here: https://github.com/Qiskit/qiskit-terra/pull/686/files#r210262725 and it can be removed.

File: test/python/common.py [hunk 44]
```diff
@@ -367,13 +370,13 @@ def _wrapper(self, *args, **kwargs):
         qiskit_wrapper._DEFAULT_PROVIDER = DefaultQISKitProvider()
         kwargs.update(_get_credentials(self, TEST_OPTIONS))
 
+        decorated_item = func
```
Comment (delapuente) [inline test/python/common.py]: I forgot to change `decorated_item` with `decorated_func`. Do you mind to do it for me?

File: test/python/test_quantumprogram.py [hunk 45]
```diff
@@ -1132,16 +1132,16 @@ def test_simulator_online_size(self, qe_token, qe_url):
         """
         backend_name = 'ibmq_qasm_simulator'
         q_program = QuantumProgram()
-        qr = q_program.create_quantum_register("q", 31)
-        cr = q_program.create_classical_register("c", 31)
+        qr = q_program.create_quantum_register("q", 1)
```
Comment (delapuente) [inline test/python/test_quantumprogram.py]: We are going to try merging #791 first then.

File: Makefile [hunk 46]
```diff
@@ -33,7 +33,7 @@ test_mock:
 	env QISKIT_TESTS=mock_online python3 -m unittest discover -s test -v
 
 test_recording:
-	rm test/cassettes/*
+	rm test/cassettes/* || true
```
Comment (delapuente) [inline Makefile]: Non-blocking, [according to Make documentation](http://www2.ipp.mpg.de/~dpc/gmake/make_47.html), have you tried prepending `-` before `rm` to ignore the errors?

File: test/python/common.py (hunk 44 above)
Comment (1ucian0) [reply test/python/common.py]: 88be95c

File: Makefile (hunk 46 above)
Comment (1ucian0) [reply Makefile]: much better c894e36

File: test/python/test_quantumprogram.py (hunk 45 above)
Comment (1ucian0) [reply test/python/test_quantumprogram.py]: #791 was solved.
